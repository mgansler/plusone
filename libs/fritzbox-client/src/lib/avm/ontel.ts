import type { Tr064 } from '../tr064/tr064'

import type { GetCallListResponse, GetPhonebookResponse } from './ontel.schemas'
import { getCallListSchema, getPhonebookSchema } from './ontel.schemas'
import { xmlParser } from './shared'

export class OnTelService {
  // https://avm.de/fileadmin/user_upload/Global/Service/Schnittstellen/x_contactSCPD.pdf
  private ServiceType = 'urn:dslforum-org:service:X_AVM-DE_OnTel:1'

  constructor(private readonly tr064: Tr064) {}

  public async getCallList(): Promise<GetCallListResponse> {
    const outVars = await this.tr064.callAction(this.ServiceType, 'GetCallList')
    const response = await this.tr064.callUrlFromActionResponse(outVars['NewCallListURL'] as string)
    return getCallListSchema.parse(xmlParser.parse(response.data))
  }

  public async getPhonebook(phonebookId = 0): Promise<GetPhonebookResponse> {
    const outVars = await this.tr064.callAction(this.ServiceType, 'GetPhonebook', { NewPhonebookID: phonebookId })
    const response = await this.tr064.callUrlFromActionResponse(outVars['NewPhonebookURL'] as string)
    return getPhonebookSchema.parse(xmlParser.parse(response.data))
  }
}
