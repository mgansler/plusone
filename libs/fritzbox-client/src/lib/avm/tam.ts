import { Agent } from 'node:https'

import axios from 'axios'
import { z } from 'zod'

import type { Tr064 } from '../tr064/tr064'

import { dateTransformSchema, xmlParser } from './shared'

const messageSchema = z
  .object({
    Index: z.number(),
    Tam: z.number(),
    Called: z.number(),
    Date: dateTransformSchema,
    Duration: z.string(),
    Inbook: z.number().transform((value) => value === 1),
    Name: z.string(),
    New: z.number().transform((value) => value === 1),
    Path: z.string(),
  })
  .transform(({ Inbook, ...rest }) => ({
    ...rest,
    inBook: Inbook,
  }))

const getMessageListSchema = z.object({
  Root: z.object({
    Message: messageSchema
      .or(z.array(messageSchema))
      .transform((messages) => (Array.isArray(messages) ? messages : [messages])),
  }),
})

export type GetMessageListResponse = z.infer<typeof getMessageListSchema>

export class TamService {
  private ServiceType = 'urn:dslforum-org:service:X_AVM-DE_TAM:1'

  constructor(private readonly tr064: Tr064) {}

  async getMessageList(): Promise<GetMessageListResponse> {
    const actionName = 'GetMessageList'
    const rawResponse = await this.tr064.makeCall(this.ServiceType, actionName, {
      NewIndex: '0',
    })
    const callUrl = rawResponse['s:Envelope']['s:Body'][`u:${actionName}Response`]['NewURL']
    const response = await axios.get(callUrl, { httpsAgent: new Agent({ rejectUnauthorized: false }) })
    return getMessageListSchema.parse(xmlParser.parse(response.data))
  }
}
