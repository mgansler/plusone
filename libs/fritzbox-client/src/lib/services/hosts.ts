import type { Tr064 } from '../tr064/tr064'

import type {
  GetGenericHostEntryOutVars,
  GetHostNumberOfEntriesOutVars,
  SpecificHostEntryByIpOutVars,
} from './hosts.vars'

export class HostsService {
  public static readonly ServiceType = 'urn:dslforum-org:service:Hosts:1'

  constructor(private readonly tr064: Tr064) {}

  public async getHostNumberOfEntries() {
    const outVars = await this.tr064.callAction<GetHostNumberOfEntriesOutVars>(
      HostsService.ServiceType,
      'GetHostNumberOfEntries',
      {},
      false,
    )
    return outVars
  }

  public async getGenericHostEntry(NewIndex: number) {
    const outVars = await this.tr064.callAction<GetGenericHostEntryOutVars>(
      HostsService.ServiceType,
      'GetGenericHostEntry',
      { NewIndex },
      false,
    )
    return outVars
  }

  public async getSpecificHostEntryByIp(NewIPAddress: string) {
    const outVars = await this.tr064.callAction<SpecificHostEntryByIpOutVars>(
      HostsService.ServiceType,
      'X_AVM-DE_GetSpecificHostEntryByIP',
      {
        NewIPAddress,
      },
    )
    return outVars
  }
}
