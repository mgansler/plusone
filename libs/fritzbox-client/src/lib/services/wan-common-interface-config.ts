import type { Tr064 } from '../tr064/tr064'

import type {
  ActiveProviderOutVars,
  OnlineMonitorOutVars,
  TotalBytesReceivedOutVars,
  TotalBytesSentOutVars,
} from './wan-common-interface-config.vars'

export class WanCommonInterfaceConfigService {
  public static ServiceType = 'urn:dslforum-org:service:WANCommonInterfaceConfig:1'

  constructor(private readonly tr064: Tr064) {}

  public async getActiveProvider() {
    const outVars = await this.tr064.callAction<ActiveProviderOutVars>(
      WanCommonInterfaceConfigService.ServiceType,
      'X_AVM-DE_GetActiveProvider',
    )
    return outVars
  }

  public async getOnlineMonitor(NewSyncGroupIndex: number) {
    const outVars = await this.tr064.callAction<OnlineMonitorOutVars>(
      WanCommonInterfaceConfigService.ServiceType,
      'X_AVM-DE_GetOnlineMonitor',
      { NewSyncGroupIndex },
    )
    return outVars
  }

  public async getTotalBytesReceived() {
    const outVars = await this.tr064.callAction<TotalBytesReceivedOutVars>(
      WanCommonInterfaceConfigService.ServiceType,
      'GetTotalBytesReceived',
    )
    return this.bytesToHuman(outVars['NewTotalBytesReceived'])
  }

  public async getTotalBytesSent() {
    const outVars = await this.tr064.callAction<TotalBytesSentOutVars>(
      WanCommonInterfaceConfigService.ServiceType,
      'GetTotalBytesSent',
    )
    return this.bytesToHuman(outVars['NewTotalBytesSent'])
  }

  private bytesToHuman(bytes: number) {
    return {
      bytes: bytes,
      kiloBytes: Math.round((bytes / 1024) * 100) / 100,
      megaBytes: Math.round((bytes / 1024 / 1024) * 100) / 100,
      gigaBytes: Math.round((bytes / 1024 / 1024 / 1024) * 100) / 100,
    }
  }
}
