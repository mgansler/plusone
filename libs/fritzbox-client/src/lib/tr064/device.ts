import type { z } from 'zod'

import type { FritzBoxConfig } from '../config'

import type { deviceSchema } from './schema'
import { Service } from './service'

export class Device {
  private services: Array<Service> = []

  constructor(
    private readonly config: FritzBoxConfig,
    private readonly device: z.infer<typeof deviceSchema>,
  ) {}

  public async parseServices() {
    for (const service of this.device.serviceList.service) {
      const s = new Service(this.config, service)
      await s.parseActions()
      this.services.push(s)
    }
  }

  public findServiceByType(serviceType: string): Service | undefined {
    return this.services.find((service) => service.isOfServiceType(serviceType))
  }
}
