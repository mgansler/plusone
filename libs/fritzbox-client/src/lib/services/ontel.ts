import type { Tr064 } from '../tr064/tr064'

import type { GetCallListResponse, GetPhonebookResponse } from './ontel.schemas'
import { getCallListSchema, getPhonebookSchema } from './ontel.schemas'
import type { CallListOutVars, PhonebookOutVars } from './ontel.vars'
import { xmlParser } from './shared'

export class OnTelService {
  // https://avm.de/fileadmin/user_upload/Global/Service/Schnittstellen/x_contactSCPD.pdf
  public static readonly ServiceType = 'urn:dslforum-org:service:X_AVM-DE_OnTel:1'

  constructor(private readonly tr064: Tr064) {}

  public async getCallList(): Promise<GetCallListResponse> {
    const outVars = await this.tr064.callAction<CallListOutVars>(OnTelService.ServiceType, 'GetCallList')
    const response = await this.tr064.callUrlFromActionResponse(outVars['NewCallListURL'])
    return getCallListSchema.parse(xmlParser.parse(response.data))
  }

  public async getPhonebook(phonebookId = 0): Promise<GetPhonebookResponse> {
    const outVars = await this.tr064.callAction<PhonebookOutVars>(OnTelService.ServiceType, 'GetPhonebook', {
      NewPhonebookID: phonebookId,
    })
    const response = await this.tr064.callUrlFromActionResponse(outVars['NewPhonebookURL'])
    return getPhonebookSchema.parse(xmlParser.parse(response.data))
  }
}
