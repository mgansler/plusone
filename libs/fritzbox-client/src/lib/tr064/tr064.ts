import { createHash } from 'node:crypto'
import { Agent } from 'node:https'

import type { AxiosInstance } from 'axios'
import axios from 'axios'

import { xmlBuilder, xmlParser } from '../avm/shared'
import type { FritzBoxConfig } from '../config'

import { tr064DescSchema } from './schema'
import { Service } from './service'

type AuthParameters = {
  Nonce: string
  Realm: string
}

export class Tr064 {
  public authParams: AuthParameters | undefined

  private constructor(
    private readonly config: FritzBoxConfig,
    private services: Array<Service> = [],
    private readonly axiosInstance: AxiosInstance,
  ) {}

  static async init(config: FritzBoxConfig): Promise<Tr064> {
    const securityPortResponse = await axios.post(
      `http://${config.host}:49000/upnp/control/deviceinfo`,
      xmlBuilder.build({
        's:Envelope': {
          '@@xmlns:s': 'http://schemas.xmlsoap.org/soap/envelope/',
          's:Body': {
            'u:GetSecurityPort': { '@@xmlns:u': 'urn:dslforum-org:service:DeviceInfo:1' },
          },
        },
      }),
      {
        headers: {
          'Content-Type': 'text/xml; charset=utf-8',
          SOAPAction: 'urn:dslforum-org:service:DeviceInfo:1#GetSecurityPort',
        },
      },
    )

    const newSecurityPort = xmlParser.parse(securityPortResponse.data)['s:Envelope']['s:Body'][
      'u:GetSecurityPortResponse'
    ]['NewSecurityPort']

    if (typeof newSecurityPort !== 'number') {
      throw new Error('Failed to retrieve security port.')
    }

    const tr64DescXMLResponse = await axios.get(`https://${config.host}:${newSecurityPort}/tr64desc.xml`, {
      httpsAgent: new Agent({
        rejectUnauthorized: false,
      }),
    })
    const tr064Desc = tr064DescSchema.parse(xmlParser.parse(tr64DescXMLResponse.data))

    const axiosInstance = axios.create({
      baseURL: `https://${config.host}:${newSecurityPort}`,
      httpsAgent: new Agent({
        rejectUnauthorized: false,
      }),
    })

    const services: Array<Service> = []
    for (const serviceDescription of tr064Desc.root.device.serviceList.service) {
      const service = new Service({ host: config.host, port: newSecurityPort }, serviceDescription)
      await service.parseActions()
      services.push(service)
    }

    return new Tr064(config, services, axiosInstance)
  }

  public findServiceByType(serviceType: string): Service | undefined {
    return this.services.find((service) => service.isOfServiceType(serviceType))
  }

  public async callAction(
    serviceType: string,
    actionName: string,
    inArgs: Record<string, string | number> = {},
    isRetry = false,
  ): Promise<Record<string, string | number>> {
    const service = this.findServiceByType(serviceType)
    if (!service) {
      throw new Error(`No service of type '${serviceType}' found.`)
    }

    const action = service.findAction(actionName)

    if (!action) {
      throw new Error(`Action named '${actionName}' does not exist in service '${serviceType}'.`)
    }

    action.forEach((arg) => {
      if (arg.direction === 'in' && !Object.keys(inArgs).includes(arg.name)) {
        throw new Error(`Missing required inArg '${arg.name}'.`)
      }
    })

    const headers = {
      'Content-Type': 'text/xml; charset=utf-8',
      SOAPAction: `${serviceType}#${actionName}`,
    }

    if (!this.authParams) {
      const resp = await this.axiosInstance.post(
        service.controlUrl,
        this.constructInitialSoapEnvelope(serviceType, actionName, inArgs),
        { headers },
      )
      const parsedData = xmlParser.parse(resp.data)
      const challenge = parsedData['s:Envelope']['s:Header']['h:Challenge']

      if (!challenge) {
        throw new Error('Initial response contained no challenge, something went wrong.')
      }

      this.authParams = { Nonce: challenge['Nonce'], Realm: challenge['Realm'] }
    }

    const respWithAuth = await this.axiosInstance.post(
      service.controlUrl,
      this.constructSoapEnvelope(serviceType, actionName, this.authParams, inArgs),
      { headers },
    )

    const parsedResponse = xmlParser.parse(respWithAuth.data)
    const challenge = parsedResponse['s:Envelope']['s:Header']['h:Challenge']
    if (challenge && challenge['Status'] === 'Unauthenticated') {
      if (isRetry) {
        this.authParams = undefined
        throw new Error('Re-Authentication failed.')
      }
      this.authParams = { Nonce: challenge['Nonce'], Realm: challenge['Realm'] }
      return this.callAction(serviceType, actionName, inArgs, true)
    }

    const actionResponse = parsedResponse['s:Envelope']['s:Body'][`u:${actionName}Response`]
    return action.reduce(
      (outVars, variable) => {
        return variable.direction !== 'out'
          ? outVars
          : {
              ...outVars,
              [variable.name]: actionResponse[variable.name],
            }
      },
      {} as Record<string, string | number>,
    )
  }

  /**
   * The returned callUrl may be in the format `https://[2001:db8::1]/xxx` which would be a valid ipv6 url.
   * Unfortunately, axios tries to resolve `[2001:db8::1]` as a hostname: https://github.com/axios/axios/issues/5333.
   * Therefor we use the hostname from the config.
   */
  public async callUrlFromActionResponse(callUrl: string) {
    const url = new URL(callUrl)
    return this.axiosInstance.get(url.pathname + url.search)
  }

  private calcAuthDigest(realm: string, nonce: string): string {
    const secret = createHash('md5').update(`${this.config.username}:${realm}:${this.config.password}`).digest('hex')
    return createHash('md5').update(`${secret}:${nonce}`).digest('hex')
  }

  private constructInitialSoapEnvelope(
    serviceType: string,
    actionName: string,
    inArgs: Record<string, string | number>,
  ): string {
    return xmlBuilder.build({
      '?xml': { '@@version': '1.0', '@@encoding': 'utf8' },
      's:Envelope': {
        '@@s:encodingStyle': 'http://schemas.xmlsoap.org/soap/encoding/',
        '@@xmlns:s': 'http://schemas.xmlsoap.org/soap/envelope/',
        's:Header': {
          'h:InitChallenge': {
            '@@xmlns:h': 'http://soap-authentication.org/digest/2001/10/',
            '@@s:mustUnderstand': '1',
            UserID: this.config.username,
          },
        },
        's:Body': {
          [`u:${actionName}`]: { '@@xmlns:u': serviceType, ...inArgs },
        },
      },
    })
  }

  private constructSoapEnvelope(
    serviceType: string,
    actionName: string,
    authParams: AuthParameters,
    inArgs: Record<string, string | number>,
  ): string {
    return xmlBuilder.build({
      '?xml': { '@@version': '1.0', '@@encoding': 'utf8' },
      's:Envelope': {
        '@@s:encodingStyle': 'http://schemas.xmlsoap.org/soap/encoding/',
        '@@xmlns:s': 'http://schemas.xmlsoap.org/soap/envelope/',
        's:Header': {
          'h:ClientAuth': {
            '@@xmlns:h': 'http://soap-authentication.org/digest/2001/10/',
            '@@s:mustUnderstand': '1',
            UserID: this.config.username,
            Nonce: authParams.Nonce,
            Auth: this.calcAuthDigest(authParams.Realm, authParams.Nonce),
            Realm: authParams.Realm,
          },
        },
        's:Body': {
          [`u:${actionName}`]: { '@@xmlns:u': serviceType, ...inArgs },
        },
      },
    })
  }
}
