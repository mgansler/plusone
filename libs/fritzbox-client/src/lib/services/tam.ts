import type { Tr064 } from '../tr064/tr064'

import { xmlParser } from './shared'
import type { GetMessageListResponse } from './tam.schema'
import { getMessageListSchema } from './tam.schema'
import type { MessageListOutVars } from './tam.vars'

export class TamService {
  public static ServiceType = 'urn:dslforum-org:service:X_AVM-DE_TAM:1'

  constructor(private readonly tr064: Tr064) {}

  public async getMessageList(): Promise<GetMessageListResponse> {
    const outVars = await this.tr064.callAction<MessageListOutVars>(TamService.ServiceType, 'GetMessageList', {
      NewIndex: '0',
    })
    const response = await this.tr064.callUrlFromActionResponse(outVars['NewURL'])
    return getMessageListSchema.parse(xmlParser.parse(response.data))
  }
}
