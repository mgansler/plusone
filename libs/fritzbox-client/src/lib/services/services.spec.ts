import { setupServer } from 'msw/node'
import { beforeEach, expect } from 'vitest'

import { deviceinfoXml } from '../stubs/deviceinfo.xml.spec'
import { hostsXmlSpec } from '../stubs/hosts.xml.spec'
import { hostsGetGenericHostEntrySpec } from '../stubs/hostsGetGenericHostEntry.spec'
import { hostsGetHostNumberOfEntriesSpec } from '../stubs/hostsGetHostNumberOfEntries.spec'
import { hostsGetSpecificHostEntrySpec } from '../stubs/hostsGetSpecificHostEntry.spec'
import { tr64descXmlSpec } from '../stubs/tr64desc.xml.spec'
import { wancommonifconfigGetActiveProviderXmlSpec } from '../stubs/wancommonifconfigGetActiveProvider.xml.spec'
import { wancommonifconfigGetOnlineMonitorXmlSpec } from '../stubs/wancommonifconfigGetOnlineMonitor.xml.spec'
import { wancommonifconfigGetTotalBytesReceivedXmlSpec } from '../stubs/wancommonifconfigGetTotalBytesReceived.xml.spec'
import { wancommonifconfigGetTotalBytesSentXmlSpec } from '../stubs/wancommonifconfigGetTotalBytesSent.xml.spec'
import { wancommonifconfigSCPDXmlSpec } from '../stubs/wancommonifconfigSCPD.xml.spec'
import {
  x_contactCallList_oneEntry,
  x_contactCallList_twoEntries,
  x_contactGetCallListXmlSpec,
} from '../stubs/x_contactGetCallList.xml.spec'
import { x_contactGetPhonebookXmlSpec, x_contactPhonebook } from '../stubs/x_contactGetPhonebook.xml.spec'
import { x_contactSCPDXmlSpec } from '../stubs/x_contactSCPD.xml.spec'
import {
  x_tamGetMessageListXmlSpec,
  x_tamMessageList_oneEntry,
  x_tamMessageList_twoEntries,
} from '../stubs/x_tamGetMessageList.xml.spec'
import { x_tamSCPDXmlSpec } from '../stubs/x_tamSCPD.xml.spec'
import { Tr064 } from '../tr064/tr064'

import { HostsService } from './hosts'
import { OnTelService } from './ontel'
import { TamService } from './tam'
import { WanCommonInterfaceConfigService } from './wan-common-interface-config'

describe('Services', () => {
  const server = setupServer(
    tr64descXmlSpec,
    deviceinfoXml,
    hostsXmlSpec,
    wancommonifconfigSCPDXmlSpec, // WanCommonInterfaceConfig
    x_contactSCPDXmlSpec, // OnTel
    x_tamSCPDXmlSpec, // TAM
  )
  let tr064: Tr064

  beforeAll(async () => {
    server.listen({ onUnhandledRequest: 'error' })
    tr064 = await Tr064.init({ host: 'fritz-test.box', username: 'admin', password: 'gurkensalat' })
  })

  afterAll(() => {
    server.close()
  })

  describe('Hosts', () => {
    let hostsService: HostsService

    beforeEach(() => {
      hostsService = new HostsService(tr064)
    })

    it('should return the number of known hosts', async () => {
      server.use(hostsGetHostNumberOfEntriesSpec)

      const numberOfHosts = await hostsService.getHostNumberOfEntries()

      expect(numberOfHosts).toMatchSnapshot()
    })

    it('should return a generic host entry', async () => {
      server.use(hostsGetGenericHostEntrySpec)

      const genericHostEntry = await hostsService.getGenericHostEntry(0)

      expect(genericHostEntry).toMatchSnapshot()
    })

    it('should return a specific host entry by ip', async () => {
      server.use(hostsGetSpecificHostEntrySpec)

      const specificHostEntry = await hostsService.getSpecificHostEntryByIp('127.0.0.1')

      expect(specificHostEntry).toMatchSnapshot()
    })
  })

  describe('X_AVM-DE_OnTel', () => {
    let onTelService: OnTelService

    beforeEach(() => {
      onTelService = new OnTelService(tr064)
    })

    describe('getCallList', () => {
      it('should return a call list with a single entry', async () => {
        server.use(x_contactGetCallListXmlSpec, x_contactCallList_oneEntry)

        const callList = await onTelService.getCallList()
        expect(callList.root.Call).toHaveLength(1)
        expect(callList.root.Call[0]).toMatchSnapshot()
      })

      it('should retrieve the call list', async () => {
        server.use(x_contactGetCallListXmlSpec, x_contactCallList_twoEntries)
        const callList = await onTelService.getCallList()

        expect(callList.root.Call).toHaveLength(2)
        expect(callList.root.Call[0]).toMatchSnapshot()
        expect(callList.root.Call[1]).toMatchSnapshot()
      })
    })

    describe('getPhonebook', () => {
      it('should get the phonebook', async () => {
        server.use(x_contactGetPhonebookXmlSpec, x_contactPhonebook)

        const phonebook = await onTelService.getPhonebook()
        expect(phonebook.phonebooks.phonebook.contact).toHaveLength(12)
        expect(phonebook.phonebooks.phonebook.contact).toMatchSnapshot()
      })
    })
  })

  describe('X_AVM-DE_TAM', () => {
    let tamService: TamService

    beforeEach(() => {
      tamService = new TamService(tr064)
    })

    it('should retrieve the message list with one entry', async () => {
      server.use(x_tamGetMessageListXmlSpec, x_tamMessageList_oneEntry)

      const messageList = await tamService.getMessageList()

      expect(messageList.Root.Message).toHaveLength(1)
      expect(messageList.Root.Message[0]).toMatchSnapshot()
    })

    it('should retrieve the message list with two entries', async () => {
      server.use(x_tamGetMessageListXmlSpec, x_tamMessageList_twoEntries)

      const messageList = await tamService.getMessageList()

      expect(messageList.Root.Message).toHaveLength(2)
      expect(messageList.Root.Message[0]).toMatchSnapshot()
      expect(messageList.Root.Message[1]).toMatchSnapshot()
    })
  })

  describe('WanCommonInterfaceConfig', () => {
    let wanCommonInterfaceConfigService: WanCommonInterfaceConfigService

    beforeEach(() => {
      wanCommonInterfaceConfigService = new WanCommonInterfaceConfigService(tr064)
    })

    it('should get the active provider', async () => {
      server.use(wancommonifconfigGetActiveProviderXmlSpec)

      const activeProvider = await wanCommonInterfaceConfigService.getActiveProvider()

      expect(activeProvider).toMatchSnapshot()
    })

    it('should get the online monitor', async () => {
      server.use(wancommonifconfigGetOnlineMonitorXmlSpec)

      const onlineMonitor = await wanCommonInterfaceConfigService.getOnlineMonitor(0)

      expect(onlineMonitor).toMatchSnapshot()
    })

    it('should get the total bytes received', async () => {
      server.use(wancommonifconfigGetTotalBytesReceivedXmlSpec)

      const totalBytesReceived = await wanCommonInterfaceConfigService.getTotalBytesReceived()

      expect(totalBytesReceived).toMatchSnapshot()
    })

    it('should get the total bytes sent', async () => {
      server.use(wancommonifconfigGetTotalBytesSentXmlSpec)

      const totalBytesSent = await wanCommonInterfaceConfigService.getTotalBytesSent()

      expect(totalBytesSent).toMatchSnapshot()
    })
  })
})
