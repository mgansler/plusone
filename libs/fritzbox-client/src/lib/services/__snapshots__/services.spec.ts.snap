// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Services > Hosts > should return a generic host entry 1`] = `
{
  "NewActive": 1,
  "NewAddressSource": "DHCP",
  "NewHostName": "fritz.box",
  "NewIPAddress": "192.168.178.1",
  "NewInterfaceType": "Ethernet",
  "NewLeaseTimeRemaining": 0,
  "NewMACAddress": "MA:CA:DD:RE:SS:00",
}
`;

exports[`Services > Hosts > should return a specific host entry by ip 1`] = `
{
  "NewActive": 1,
  "NewHostName": "fritz.box",
  "NewInterfaceType": "",
  "NewMACAddress": "MA:CA:DD:RE:SS:00",
  "NewX_AVM-DE_Disallow": 0,
  "NewX_AVM-DE_FriendlyName": "fritz.box",
  "NewX_AVM-DE_FriendlyNameIsWriteable": 0,
  "NewX_AVM-DE_Guest": 0,
  "NewX_AVM-DE_InfoURL": "",
  "NewX_AVM-DE_IsMeshable": 1,
  "NewX_AVM-DE_MACAddressList": "MA:CA:DD:RE:SS:00,MA:CA:DD:RE:SS:01",
  "NewX_AVM-DE_Model": "",
  "NewX_AVM-DE_Port": 0,
  "NewX_AVM-DE_Priority": 0,
  "NewX_AVM-DE_RequestClient": 0,
  "NewX_AVM-DE_Speed": 0,
  "NewX_AVM-DE_URL": "http://192.168.178.1",
  "NewX_AVM-DE_UpdateAvailable": 0,
  "NewX_AVM-DE_UpdateSuccessful": "unknown",
  "NewX_AVM-DE_VPN": 0,
  "NewX_AVM-DE_WANAccess": "granted",
}
`;

exports[`Services > Hosts > should return the number of known hosts 1`] = `
{
  "NewHostNumberOfEntries": 42,
}
`;

exports[`Services > WanCommonInterfaceConfig > should get the active provider 1`] = `
{
  "NewX_AVM-DE_Provider": "Vodafone",
}
`;

exports[`Services > WanCommonInterfaceConfig > should get the online monitor 1`] = `
{
  "NewSyncGroupMode": "CABLE",
  "NewSyncGroupName": "sync_cable",
  "NewTotalNumberSyncGroups": 2,
  "Newds_current_bps": "1711,2287,14023,2291,892,27713,27713,896,10582,2333,25519,15651,5088,6102,4115,2051,3267,1415,9443,3255",
  "Newmax_ds": 857938957,
  "Newmax_us": 279624825,
  "Newmc_current_bps": "0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0",
  "Newprio_default_bps": "763,767,691,332,719,857,2044,1102,859,693,652,651,633,2135,1700,469,771,825,519,719",
  "Newprio_high_bps": "0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0",
  "Newprio_low_bps": "0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0",
  "Newprio_realtime_bps": "396,15174,20835,6281,6281,1478,6587,16180,9765,778,7097,20764,5322,6607,3596,4360,5676,1107,161,4966",
  "Newus_current_bps": "1159,15941,21526,6613,7000,2335,8631,17282,10624,1471,7749,21415,5955,8742,5296,4829,6447,1932,680,5685",
}
`;

exports[`Services > WanCommonInterfaceConfig > should get the total bytes received 1`] = `
{
  "bytes": 2267246611,
  "gigaBytes": 2.11,
  "kiloBytes": 2214108.02,
  "megaBytes": 2162.21,
}
`;

exports[`Services > WanCommonInterfaceConfig > should get the total bytes sent 1`] = `
{
  "bytes": 2156315275,
  "gigaBytes": 2.01,
  "kiloBytes": 2105776.64,
  "megaBytes": 2056.42,
}
`;

exports[`Services > X_AVM-DE_OnTel > getCallList > should retrieve the call list 1`] = `
{
  "Called": "SIP: 123456789",
  "Caller": 987654321,
  "Count": "",
  "Date": 2000-12-23T23:00:00.000Z,
  "Device": "FRITZ!Fon C5",
  "Duration": "0:05",
  "Id": 2,
  "Name": "Somebody I know",
  "Numbertype": "sip",
  "Path": "",
  "Port": 10,
  "Type": 1,
}
`;

exports[`Services > X_AVM-DE_OnTel > getCallList > should retrieve the call list 2`] = `
{
  "Called": "SIP: 123456789",
  "Caller": 987654321,
  "Count": "",
  "Date": 2000-12-22T23:00:00.000Z,
  "Device": "",
  "Duration": "0:00",
  "Id": 1,
  "Name": "",
  "Numbertype": "sip",
  "Path": "",
  "Port": -1,
  "Type": 2,
}
`;

exports[`Services > X_AVM-DE_OnTel > getCallList > should return a call list with a single entry 1`] = `
{
  "Called": "SIP: 123456789",
  "Caller": 987654321,
  "Count": "",
  "Date": 2000-12-23T23:00:00.000Z,
  "Device": "FRITZ!Fon C5",
  "Duration": "0:05",
  "Id": 1,
  "Name": "Somebody I know",
  "Numbertype": "sip",
  "Path": "",
  "Port": 10,
  "Type": 1,
}
`;

exports[`Services > X_AVM-DE_OnTel > getPhonebook > should get the phonebook 1`] = `
[
  {
    "category": 0,
    "person": {
      "realName": "Alarmclock 1",
    },
    "telephony": {
      "numbers": [
        {
          "prio": "1",
          "type": "",
          "value": "**41",
        },
      ],
      "services": "",
    },
    "uniqueid": 5,
  },
  {
    "category": 0,
    "person": {
      "realName": "Alarmclock 2",
    },
    "telephony": {
      "numbers": [
        {
          "prio": "1",
          "type": "",
          "value": "**42",
        },
      ],
      "services": "",
    },
    "uniqueid": 6,
  },
  {
    "category": 0,
    "person": {
      "realName": "Alarmclock 3",
    },
    "telephony": {
      "numbers": [
        {
          "prio": "0",
          "type": "",
          "value": "**43",
        },
      ],
      "services": "",
    },
    "uniqueid": 7,
  },
  {
    "category": 0,
    "person": {
      "realName": "All internal",
    },
    "telephony": {
      "numbers": [
        {
          "prio": "1",
          "type": "intern",
          "value": "**9",
        },
      ],
      "services": "",
    },
    "uniqueid": 3,
  },
  {
    "category": 0,
    "person": {
      "realName": "Answering Machine",
    },
    "telephony": {
      "numbers": [
        {
          "prio": "1",
          "type": "intern",
          "value": "**600",
        },
        {
          "prio": "0",
          "type": "memo",
          "value": "**605",
        },
      ],
      "services": "",
    },
    "uniqueid": 11,
  },
  {
    "category": 0,
    "person": {
      "realName": "AVM voice clip (HD)",
    },
    "telephony": {
      "numbers": [
        {
          "prio": "1",
          "type": "work",
          "value": "550@hd-telefonie.avm.de",
        },
      ],
      "services": "",
    },
    "uniqueid": 0,
  },
  {
    "category": 0,
    "person": {
      "realName": "DECT all",
    },
    "telephony": {
      "numbers": [
        {
          "prio": "1",
          "type": "",
          "value": "**50",
        },
      ],
      "services": "",
    },
    "uniqueid": 4,
  },
  {
    "category": 0,
    "person": {
      "realName": "FRITZ!Fon C5",
    },
    "telephony": {
      "numbers": [
        {
          "prio": "1",
          "type": "intern",
          "value": "**610",
        },
      ],
      "services": "",
    },
    "uniqueid": 12,
  },
  {
    "category": 1,
    "person": {
      "realName": "Some important person",
    },
    "telephony": {
      "numbers": [
        {
          "prio": "1",
          "type": "home",
          "value": 123456789,
        },
      ],
      "services": "",
    },
    "uniqueid": 13,
  },
  {
    "category": 0,
    "person": {
      "realName": "A not so important person",
    },
    "telephony": {
      "numbers": [
        {
          "prio": "1",
          "type": "work",
          "value": 987654321,
        },
      ],
      "services": "",
    },
    "uniqueid": 14,
  },
  {
    "category": 0,
    "person": {
      "realName": "Telephone",
    },
    "telephony": {
      "numbers": [
        {
          "prio": "1",
          "type": "intern",
          "value": "**1",
        },
      ],
      "services": "",
    },
    "uniqueid": 8,
  },
  {
    "category": 0,
    "person": {
      "realName": "Telephone",
    },
    "telephony": {
      "numbers": [
        {
          "prio": "1",
          "type": "intern",
          "value": "**2",
        },
      ],
      "services": "",
    },
    "uniqueid": 9,
  },
]
`;

exports[`Services > X_AVM-DE_TAM > should retrieve the message list with one entry 1`] = `
{
  "Called": 123456789,
  "Date": 2000-12-22T23:00:00.000Z,
  "Duration": "0:01",
  "Index": 0,
  "Name": "Somebody I know",
  "New": false,
  "Path": "/download.lua?path=/data/tam/rec/rec.0.000",
  "Tam": 0,
  "inBook": true,
}
`;

exports[`Services > X_AVM-DE_TAM > should retrieve the message list with two entries 1`] = `
{
  "Called": 123456789,
  "Date": 2000-12-23T23:00:00.000Z,
  "Duration": "0:01",
  "Index": 1,
  "Name": "",
  "New": true,
  "Path": "/download.lua?path=/data/tam/rec/rec.0.001",
  "Tam": 0,
  "inBook": false,
}
`;

exports[`Services > X_AVM-DE_TAM > should retrieve the message list with two entries 2`] = `
{
  "Called": 123456789,
  "Date": 2000-12-22T23:00:00.000Z,
  "Duration": "0:01",
  "Index": 0,
  "Name": "Somebody I know",
  "New": false,
  "Path": "/download.lua?path=/data/tam/rec/rec.0.000",
  "Tam": 0,
  "inBook": true,
}
`;
